File: bindings/ruby/test/fossh_test.rb
# frozen_string_literal: true

# Exercises the accept/reject/rate-limit/wrong-key paths end to end
# against a real database, using the `fossh` CLI (assumed built and on
# $PATH, or reachable via $FOSSH_CLI_BIN) to `init`/`site create` a
# throwaway environment — the same pattern bindings/go/fossh_test.go
# uses, just driven from Ruby.
#
# Requires `libfossh.so` findable via `$FOSSH_LIB_PATH` (or already on
# the loader's default search path) — not runnable in an environment
# without a Ruby interpreter, which is the state this was authored in;
# see DECISIONS.md for that limitation.

require "minitest/autorun"
require "English"
require "fileutils"
require "tmpdir"
require_relative "../lib/fossh"

class FosshIntegrationTest < Minitest::Test
  def cli_bin
    bin = ENV["FOSSH_CLI_BIN"]
    return bin if bin && !bin.empty?

    path = ENV["PATH"].split(File::PATH_SEPARATOR).map { |dir| File.join(dir, "fossh") }.find { |p| File.executable?(p) }
    skip "fossh CLI binary not found on $PATH and $FOSSH_CLI_BIN not set; skipping integration test" unless path
    path
  end

  def setup_site(allow: [])
    bin = cli_bin
    dir = Dir.mktmpdir("fossh-ruby-test")
    data_dir = File.join(dir, "data")

    init_out = `#{bin} init --dir #{data_dir} 2>&1`
    raise "fossh init failed:\n#{init_out}" unless $CHILD_STATUS.success?

    config_path = File.join(dir, "fossh.toml")
    contents = File.read(config_path)
    patched = contents.sub('mode = "spool"', 'mode = "direct"')
    patched += "\n[rate_limit]\nper_sec = 2\nburst = 2\n"
    File.write(config_path, patched)
    File.chmod(0o600, config_path)

    args = ["site", "create", "testsite"]
    args += ["--allow", allow.join(",")] unless allow.empty?
    out = IO.popen({ "FOSSH_CONFIG" => config_path }, [bin] + args, err: %i[child out], &:read)
    raise "fossh site create failed:\n#{out}" unless $CHILD_STATUS.success?

    write_key = out.lines.map(&:strip).find { |line| line.start_with?("fossh_testsite_") }
    raise "could not find write key in output:\n#{out}" unless write_key

    [config_path, write_key]
  end

  def test_pageview_and_event_accept_and_reject
    config_path, key = setup_site(allow: %w[pageview signup])
    client = FoSSH::Client.new(config: config_path, key: key)

    assert client.event("signup", value: 1)

    err = assert_raises(FoSSH::Error) { client.timing("db.query", 0) }
    assert_equal "REJECTED", err.name, "an unallowlisted name must be rejected"
  ensure
    client&.close
  end

  def test_event_with_allowlisted_prop
    config_path, key = setup_site(allow: %w[checkout tier])
    client = FoSSH::Client.new(config: config_path, key: key)

    assert client.event("checkout", value: 1, props: { "tier" => "pro" })
  ensure
    client&.close
  end

  def test_rate_limiting
    config_path, key = setup_site(allow: %w[pageview])
    client = FoSSH::Client.new(config: config_path, key: key)

    # burst = 2 (patched into the config in setup_site).
    assert client.event("pageview", value: 1)
    assert client.event("pageview", value: 1)

    err = assert_raises(FoSSH::Error) { client.event("pageview", value: 1) }
    assert_equal "RATE_LIMITED", err.name
  ensure
    client&.close
  end

  def test_wrong_key_is_unauthorized
    config_path, = setup_site(allow: %w[pageview])
    wrong_key = "fossh_testsite_#{"A" * 52}"

    # fossh_set_key failing during construction surfaces as FoSSH::Error
    # (a real, loaded fossh_ctx that rejected the key) — distinct from
    # FoSSH::LoadError, which is specifically "the library/context
    # itself never came up at all."
    err = assert_raises(FoSSH::Error) do
      FoSSH::Client.new(config: config_path, key: wrong_key)
    end
    assert_equal "UNAUTHORIZED", err.name
  end
end

File: bindings/go/fossh_test.go
//go:build !nofossh

package fossh_test

// Exercises the accept/reject/rate-limit paths end to end against a real
// database, using the `fossh` CLI (assumed built and on $PATH, or
// reachable via $FOSSH_CLI_BIN) to `init`/`site create` a throwaway
// environment — the same setup this whole repository's other
// integration tests use, just driven from Go instead of shelling out
// from a Rust test.
//
// Requires CGO_CFLAGS/CGO_LDFLAGS pointed at this repo's include/ and a
// built libfossh (see docs/INTEGRATION-go.md) — not runnable in an
// environment without a Go toolchain, which is the state this was
// authored in; see DECISIONS.md for that limitation.

import (
	"encoding/json"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"testing"

	fossh "github.com/s0aptile/fossh-go"
)

func cliBin(t *testing.T) string {
	t.Helper()
	if bin := os.Getenv("FOSSH_CLI_BIN"); bin != "" {
		return bin
	}
	bin, err := exec.LookPath("fossh")
	if err != nil {
		t.Skip("fossh CLI binary not found on $PATH and $FOSSH_CLI_BIN not set; skipping integration test")
	}
	return bin
}

type testSite struct {
	configPath string
	writeKey   string
}

func setupSite(t *testing.T, allow []string) testSite {
	t.Helper()
	bin := cliBin(t)
	dir := t.TempDir()
	dataDir := filepath.Join(dir, "data")

	initCmd := exec.Command(bin, "init", "--dir", dataDir)
	initCmd.Dir = dir
	if out, err := initCmd.CombinedOutput(); err != nil {
		t.Fatalf("fossh init: %v\n%s", err, out)
	}

	configPath := filepath.Join(dir, "fossh.toml")
	// Direct mode + a tiny burst so the rate-limit test doesn't need
	// hundreds of calls to exhaust it.
	contents, err := os.ReadFile(configPath)
	if err != nil {
		t.Fatalf("reading generated fossh.toml: %v", err)
	}
	patched := strings.Replace(string(contents), `mode = "spool"`, `mode = "direct"`, 1)
	patched += "\n[rate_limit]\nper_sec = 2\nburst = 2\n"
	if err := os.WriteFile(configPath, []byte(patched), 0o600); err != nil {
		t.Fatalf("patching fossh.toml: %v", err)
	}

	args := []string{"site", "create", "testsite"}
	if len(allow) > 0 {
		args = append(args, "--allow", strings.Join(allow, ","))
	}
	createCmd := exec.Command(bin, args...)
	createCmd.Env = append(os.Environ(), "FOSSH_CONFIG="+configPath)
	out, err := createCmd.CombinedOutput()
	if err != nil {
		t.Fatalf("fossh site create: %v\n%s", err, out)
	}

	var writeKey string
	for _, line := range strings.Split(string(out), "\n") {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "fossh_testsite_") {
			writeKey = line
			break
		}
	}
	if writeKey == "" {
		t.Fatalf("could not find write key in output:\n%s", out)
	}

	return testSite{configPath: configPath, writeKey: writeKey}
}

func TestPageviewAndEventAccept(t *testing.T) {
	site := setupSite(t, []string{"pageview", "signup"})
	client, err := fossh.New(fossh.Config{Path: site.configPath, Key: site.writeKey})
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	defer client.Close()

	if err := client.Event("signup", 1, nil); err != nil {
		t.Errorf("Event(signup): %v", err)
	}
	if err := client.Timing("db.query", 0); err == nil {
		t.Errorf("Timing(db.query) with an unallowlisted name should have been rejected")
	} else if fe, ok := err.(*fossh.Error); !ok || fe.Name != "REJECTED" {
		t.Errorf("expected a REJECTED *fossh.Error, got %v (%T)", err, err)
	}
}

func TestEventWithProps(t *testing.T) {
	site := setupSite(t, []string{"checkout", "tier"})
	client, err := fossh.New(fossh.Config{Path: site.configPath, Key: site.writeKey})
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	defer client.Close()

	if err := client.Event("checkout", 1, map[string]string{"tier": "pro"}); err != nil {
		t.Errorf("Event with allowlisted prop: %v", err)
	}
}

func TestRateLimiting(t *testing.T) {
	site := setupSite(t, []string{"pageview"})
	client, err := fossh.New(fossh.Config{Path: site.configPath, Key: site.writeKey})
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	defer client.Close()

	// burst = 2 (patched into the config in setupSite).
	if err := client.Event("pageview", 1, nil); err != nil {
		t.Fatalf("first event within burst: %v", err)
	}
	if err := client.Event("pageview", 1, nil); err != nil {
		t.Fatalf("second event within burst: %v", err)
	}
	err = client.Event("pageview", 1, nil)
	if err == nil {
		t.Fatal("third event should have exceeded the burst of 2")
	}
	if fe, ok := err.(*fossh.Error); !ok || fe.Name != "RATE_LIMITED" {
		t.Errorf("expected a RATE_LIMITED *fossh.Error, got %v (%T)", err, err)
	}
}

func TestWrongKeyIsUnauthorized(t *testing.T) {
	site := setupSite(t, []string{"pageview"})
	_, err := fossh.New(fossh.Config{Path: site.configPath, Key: "fossh_testsite_" + strings.Repeat("A", 52)})
	if err == nil {
		t.Fatal("expected New to fail with a wrong key")
	}
	if fe, ok := err.(*fossh.Error); !ok || fe.Name != "UNAUTHORIZED" {
		t.Errorf("expected an UNAUTHORIZED *fossh.Error, got %v (%T)", err, err)
	}
}

func TestPropsMarshalAsJSONObjectOfStrings(t *testing.T) {
	// Sanity check on our own assumption about the wire shape fossh_event
	// expects for props_json — not calling into libfossh here.
	b, err := json.Marshal(map[string]string{"a": "b"})
	if err != nil {
		t.Fatal(err)
	}
	if string(b) != `{"a":"b"}` {
		t.Errorf("unexpected JSON shape: %s", b)
	}
}
